examples/systemcalls/systemcalls.c

```diff
@@ -93,6 +141,38 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *
 */
 
+pid_t ret_value=fork();
+int val;
+if(ret_value==-1)            //Child process is unsuccessful
+{
+	exit (-1);
+}
+else if(ret_value==0)        //Child process is successful
+{
+	int fd = open(outputfile, O_WRONLY|O_TRUNC|O_CREAT, 0644);  //open() system call opens the file specified by pathname
+	if (fd < 0) { perror("open"); abort(); } //file open failed
+	if (dup2(fd, 1) < 0) { perror("dup2"); abort(); } //dup() system call allocates a new file
+	close(fd);
+	if(execv(command[0],command)==-1) //child execution fail
+	{
+		exit (-1);
+	}
+}		
+else{
+	if(waitpid(ret_value,&val,0)==-1) //Check child process pid argument has changed the state or not
+	{
+		return false;
+		}
+	if(WIFEXITED(val))               //Here child process is terminated normally
+		{
+			if(WEXITSTATUS(val)!=0) //Here process exit failed
+			{
+				return false;
```

Comment:
Missing cleanup in these return paths. 
Also see discussion in class about avoiding duplication of logic on these two very similar functions.